### PR TITLE
Product card

### PR DESCRIPTION
**Pull request by SebaKas21**

### CODE REVIEW

**Comment by dawiddworak88** [comment]
Hi @SebaKas21 , would you remove the conflicts here?

File: be/src/Infrastructure/Localization/Foundation.Localization/Resources/GlobalResources.en.resx [hunk 1]
```diff
@@ -672,6 +672,36 @@
...
     <value>Content</value>
   </data>
+  <data name="ProductCard" xml:space="preserve">
+    <value>Product card</value>
+  </data>
+  <data name="ProductCards" xml:space="preserve">
+    <value>Product cards</value>
+  </data>
+  <data name="Definition" xml:space="preserve">
+    <value>Definicja</value>
```
Comment (dawiddworak88) [inline be/src/Infrastructure/Localization/Foundation.Localization/Resources/GlobalResources.en.resx]: **suggestion:** Definition

File: be/src/Infrastructure/Localization/Foundation.Localization/Resources/ProductResources.de.resx [hunk 2]
```diff
@@ -213,6 +228,12 @@
   <data name="ProductAttributeSavedSuccessfully" xml:space="preserve">
     <value>Das Produktattribut wurde erfolgreich gespeichert.</value>
   </data>
+  <data name="ProductCardDefaultInputName" xml:space="preserve">
+    <value>NeuerGegenstand</value>
```
Comment (dawiddworak88) [inline be/src/Infrastructure/Localization/Foundation.Localization/Resources/ProductResources.de.resx]: **suggestion:** Neuer Gegenstand

File: be/src/Infrastructure/Localization/Foundation.Localization/Resources/ProductResources.en.resx [hunk 3]
```diff
@@ -180,6 +186,12 @@
   <data name="NewProductAttributeItem" xml:space="preserve">
     <value>New product attribute item</value>
   </data>
+  <data name="NewProductCard" xml:space="preserve">
+    <value>Add new product card</value>
```
Comment (dawiddworak88) [inline be/src/Infrastructure/Localization/Foundation.Localization/Resources/ProductResources.en.resx]: **suggestion:** Add a new product card

File: be/src/Infrastructure/Localization/Foundation.Localization/Resources/ProductResources.pl.resx [hunk 4]
```diff
@@ -213,6 +228,12 @@
   <data name="ProductAttributeSavedSuccessfully" xml:space="preserve">
     <value>Atrybut produktu został pomyślnie zapisany.</value>
   </data>
+  <data name="ProductCardDefaultInputName" xml:space="preserve">
+    <value>NowyElement</value>
```
Comment (dawiddworak88) [inline be/src/Infrastructure/Localization/Foundation.Localization/Resources/ProductResources.pl.resx]: **suggestion:** Nowy element

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/CategoriesController.cs [hunk 5]
```diff
@@ -22,6 +23,7 @@ public CategoriesController(IAsyncComponentModelBuilder<ComponentModelBase, Cate
 
         public async Task<IActionResult> Index()
         {
+            Console.WriteLine("TEEEEESTYY");
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/CategoriesController.cs]: **suggestion:** :)

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardController.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+using Seller.Web.Areas.Products.ViewModels;
+using System.Linq;
+using Foundation.Account.Definitions;
+
+namespace Seller.Web.Areas.Products.Controllers
+{
+    [Area("Products")]
+    public class ProductCardController : BaseController
+    {
+        private readonly IAsyncComponentModelBuilder<ComponentModelBase, ProductCardPageViewModel> productCardPageModelBuilder;
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardController.cs]: **suggestion:** _productCardPageModelBuilder

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardController.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        {
+            this.productCardPageModelBuilder = productCardPageModelBuilder;
+        }
+
+        public async Task<IActionResult> Edit(Guid? id)
+        {
+            var componentModel = new ComponentModelBase
+            {
+                Id = id,
+                IsAuthenticated = this.User.Identity.IsAuthenticated,
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardController.cs]: **suggestion:** this redundant

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardController.cs [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+                Name = this.User.Identity.Name,
+                Language = CultureInfo.CurrentUICulture.Name,
+                Token = await HttpContext.GetTokenAsync(ApiExtensionsConstants.TokenName),
+                SellerId = GuidHelper.ParseNullable((this.User.Identity as ClaimsIdentity).Claims.FirstOrDefault(x => x.Type == AccountConstants.Claims.OrganisationIdClaim)?.Value)
+            };
+
+            var viewModel = await this.productCardPageModelBuilder.BuildModelAsync(componentModel);
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardController.cs]: **suggestion (non-blocking):** _productCardPageModelBuilder

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardsController.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+
+namespace Seller.Web.Areas.Products.Controllers
+{
+    [Area("Products")]
+    public class ProductCardsController : BaseController
+    {
+        private readonly IAsyncComponentModelBuilder<ComponentModelBase, ProductCardsPageViewModel> productCardsPageModelBuilder;
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardsController.cs]: **suggestion:** _productCardsPageModelBuilder

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardsController.cs [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+        public ProductCardsController(IAsyncComponentModelBuilder<ComponentModelBase, ProductCardsPageViewModel> productCardsPageModelBuilder)
+        {
+            this.productCardsPageModelBuilder = productCardsPageModelBuilder;
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/Controllers/ProductCardsController.cs]: **suggestion:** this redundant

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/ModelBuilders/ProductCardPageModelBuilder.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        private readonly IAsyncComponentModelBuilder<ComponentModelBase, ProductCardFormViewModel> productCardFormModelBuilder;
+        private readonly IModelBuilder<FooterViewModel> footerModelBuilder;
+
+        public ProductCardPageModelBuilder(
+            IAsyncComponentModelBuilder<ComponentModelBase, SellerHeaderViewModel> headerModelBuilder,
+            IModelBuilder<MenuTilesViewModel> menuTilesModelBuilder,
+            IAsyncComponentModelBuilder<ComponentModelBase, ProductCardFormViewModel> productCardFormModelBuilder,
+            IModelBuilder<FooterViewModel> footerModelBuilder)
+        {
+            this.headerModelBuilder = headerModelBuilder;
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/ModelBuilders/ProductCardPageModelBuilder.cs]: **suggestion:** this redundant

File: be/src/Project/Web/Seller/Seller.Web/Areas/Products/ModelBuilders/ProductCardPageModelBuilder.cs [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+using Foundation.PageContent.MenuTiles.ViewModels;
+using Seller.Web.Areas.Products.ViewModels;
+using Seller.Web.Shared.ViewModels;
+using System.Threading.Tasks;
+
+namespace Seller.Web.Areas.Products.ModelBuilders
+{
+    public class ProductCardPageModelBuilder : IAsyncComponentModelBuilder<ComponentModelBase, ProductCardPageViewModel>
+    {
+        private readonly IAsyncComponentModelBuilder<ComponentModelBase, SellerHeaderViewModel> headerModelBuilder;
```
Comment (dawiddworak88) [inline be/src/Project/Web/Seller/Seller.Web/Areas/Products/ModelBuilders/ProductCardPageModelBuilder.cs]: **suggestion:** _headerModelBuilder

File: fe/projects/Seller.Portal/src/areas/Products/components/ProductCardForm/ProductCardForm.js [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+
+        const element = newElements[index];
+
+        if (element.definitionId !== undefined) {
+            const definitionItems = newElements.filter((item) => item.definitionId === element.definitionId);
+
+            if (!(definitionItems.length > 1)) {
+                const newDefinitions = definitions;
+
+                delete newDefinitions[element.definitionId];
```
Comment (dawiddworak88) [inline fe/projects/Seller.Portal/src/areas/Products/components/ProductCardForm/ProductCardForm.js]: **suggestion:** delete is dangerous, wouldn't we be able to use immutable technique to do it

File: fe/projects/Seller.Portal/src/areas/Products/components/ProductCardModal/ProductCardModal.js [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+            ...attribute, [name]: value
+        }))
+    }
+
+    const handleDefinition = () => {
+        if (productAttribute.definitionId != undefined) {
+            const currentDefinition = props.labels.definitionsOptions.find((definition) => definition.id === productAttribute.definitionId);
+
+            if (currentDefinition != null) {
+                NavigationHelper.redirect(currentDefinition.url, "_blank");
```
Comment (dawiddworak88) [inline fe/projects/Seller.Portal/src/areas/Products/components/ProductCardModal/ProductCardModal.js]: **suggestion:** I'd create a new function in NavigationHelper.openInNewTab

File: fe/projects/Seller.Portal/src/areas/Products/components/ProductCardModal/ProductCardModal.js [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+            }
+        }
+    }
+
+    const handleDisplayName = (e) => {
+        updateAttribute(e);
+        
+        const format = new RegExp(/(.|\s)*\S(.|\s)*/)
```
Comment (dawiddworak88) [inline fe/projects/Seller.Portal/src/areas/Products/components/ProductCardModal/ProductCardModal.js]: **suggestion:** Let's move it to constants and add a descriptive name what this regex represents

**Review by dawiddworak88 — CHANGES_REQUESTED** [review]

**Comment by dawiddworak88** [comment]
Hi @SebaKas21, if you could resolve the conflicts and test it in detail once again, we'd merge it and deploy to production.

**Comment by SebaKas21** [comment]
Hi @dawiddworak88 

I added minor corrections, thanks to which I eliminated errors and the product card now works properly

**Review by dawiddworak88 — APPROVED** [review]
Thank you @SebaKas21, let's go for it

**Pull request merged**